File: jenjin-core/src/main/java/com/binarymonks/jj/render/RenderGraph.java
package com.binarymonks.jj.render;

import com.badlogic.gdx.utils.ObjectMap;


public class RenderGraph {

    /**
     * render the layers in order
     */
    public ObjectMap<Integer, RenderLayer> renderLayers = new ObjectMap<>();


    public void add(String thingPath, int thingID, ObjectMap<Integer, ThingLayer> thingLayers) {
        for (ObjectMap.Entry<Integer, ThingLayer> layer : thingLayers) {
                if (!renderLayers.containsKey(layer.key)) {
                    renderLayers.put(layer.key, new RenderLayer());
                }
                renderLayers.get(layer.key).add(thingPath, thingID, layer.value);
        }
    }

    public void remove(String path, int id, ObjectMap<Integer, ThingLayer> thingLayers) {
        for (ObjectMap.Entry<Integer, ThingLayer> layer : thingLayers) {
            if (renderLayers.containsKey(layer.key)) {
                renderLayers.get(layer.key).remove(path, id);
            }
        }
    }

    /**
     * We render the ThingLayer of each Thing in batches by their ThingSpec path.
     * This will maximise the advantages of SpriteBatch.
     */
    public class RenderLayer {
        public ObjectMap<String, ObjectMap<Integer, ThingLayer>> thingLayersByThingPathAndID = new ObjectMap<>();

        public void add(String thingPath, int thingID, ThingLayer thingLayer) {
            if (!thingLayersByThingPathAndID.containsKey(thingPath)) {
                thingLayersByThingPathAndID.put(thingPath, new ObjectMap<>());
            }
            thingLayersByThingPathAndID.get(thingPath).put(thingID, thingLayer);
        }

        public void remove(String path, int id) {
            if (thingLayersByThingPathAndID.containsKey(path)) {
                thingLayersByThingPathAndID.get(path).remove(id);
            }
        }
    }
}

File: jenjin-core/src/main/kotlin/com/binarymonks/jj/core/render/RenderGraph.kt
package com.binarymonks.jj.core.render

import com.badlogic.gdx.utils.ObjectMap


class RenderGraph {

    /**
     * render the layers in order
     */
    var graphLayers = ObjectMap<Int, GraphLayer>()


    /**
     * Add a [com.binarymonks.jj.core.scenes.Scene]s render nodes to the graph
     *
     * @param specID The [com.binarymonks.jj.core.render.RenderRoot.specID]
     * @param sceneID the [com.binarymonks.jj.core.scenes.Scene.id]
     */
    fun add(specID: Int, sceneID: Int, sceneLayers: ObjectMap<Int, RenderLayer>) {
        for (layer in sceneLayers) {
            if (!graphLayers.containsKey(layer.key)) {
                graphLayers.put(layer.key, GraphLayer())
            }
            graphLayers.get(layer.key).add(specID, sceneID, layer.value)
        }
    }

    fun remove(specID: Int, id: Int, sceneLayers: ObjectMap<Int, RenderLayer>) {
        for (layer in sceneLayers) {
            if (graphLayers.containsKey(layer.key)) {
                graphLayers.get(layer.key).remove(specID, id)
            }
        }
    }

    /**
     * We render the RenderLayer of each Scene in batches by their SceneSpec id.
     * This will maximise the advantages of SpriteBatch.
     */
    inner class GraphLayer {
        var sceneLayersByScenePathAndID = ObjectMap<Int, ObjectMap<Int, RenderLayer>>()

        fun add(specID: Int, sceneID: Int, sceneLayer: RenderLayer) {
            if (!sceneLayersByScenePathAndID.containsKey(specID)) {
                sceneLayersByScenePathAndID.put(specID, ObjectMap())
            }
            sceneLayersByScenePathAndID.get(specID).put(sceneID, sceneLayer)
        }

        fun remove(specID: Int, id: Int) {
            if (sceneLayersByScenePathAndID.containsKey(specID)) {
                sceneLayersByScenePathAndID.get(specID).remove(id)
            }
        }
    }
}
